sm_c/thread.c

```diff
@@ -5,6 +5,36 @@
 #include "thread.h"
 #include "mtrap.h"
 
+
+void switch_vector_enclave(){
+  extern void trap_vector_enclave();
+  write_csr(mtvec, &trap_vector_enclave);
+}
+
+void switch_vector_host(){
+  extern void trap_vector();
+  write_csr(mtvec, &trap_vector);
+}
+
+uint64_t getRTC(){
+	return *mtime; 
+}
+
+void swap_prev_mpp(struct thread_state* thread, uintptr_t* regs){
+  //Time interrupts can occur in either user mode or supervisor mode
+
+  int old_mpp = thread->prev_mpp;
+  int mstatus = read_csr(mstatus);
+  thread->prev_mpp = mstatus & 0x800;
+
+  if(old_mpp != 0) {
```

Comment:
I did not understand this part, why do you check (old_mpp != 0) ?
Why can't we just overwrite the MPP field of the mstatus with prev_mpp?
e.g.,  
new_mstatus = (mstatus & ~0x800) | thread->prev_mpp; 
csr_write(mstatus, new_mstatus);

Reply:
Yup, this would work. Yours is much more succint, so I changed it to this.